### ISSUE TITLE
README needs info for running with ROCm on AMD GPUs

### ISSUE THREAD

**Issue by almereyda** [FoldingAtHome/containers#3]
These containers require a `--gpu` flag for Docker that is not available in a fresh install on Ubuntu 20.04 `amd64` with the `amdgpu` driver. Also the readme of the container explicitly mentions to use an `nvidia-container-runtime`. This appears a too-strict dependency, which brings vendor lock-in and a decrease of diversity in computational ecosystems.

Does it appear valuable to also support the AMD ROCm platform, or are there any plans yet to do so?

Few resources become interesting upon that sight:

- https://rocm-documentation.readthedocs.io/en/latest/Installation_Guide/Installation-Guide.html
- https://rocm-documentation.readthedocs.io/en/latest/ROCm_Virtualization_Containers/ROCm-Virtualization-&-Containers.html#rocm-docker
- https://github.com/RadeonOpenCompute/ROCm-docker
  - https://github.com/RadeonOpenCompute/ROCm-docker/blob/master/quick-start.md

* https://github.com/RadeonOpenCompute/ROCR-Runtime
* https://github.com/abuccts/rocm-container-runtime

**Comment by beberg** [comment]
Yes plan was always one fah-gpu container for all GPUs. The container has OpenCL inside, so should already work for NVIDIA and AMD GPUs.

But running the container needs the AMD drivers, and ROCm container runtime, and different docker commands.

There are two places in the README that need a "#### For NVIDIA GPUs ... #### For AMD GPUs..." subheadings. 
1. Technical Requirements: the driver version and runtime requirements are different, but everyone needs docker 19.03+ and storage.
2. Start Folding on Single Machine: the docker command flags are different.

I don't have an AMD GPU in a Linux box, so it would be great if someone could test the exact docker command, and submit the needed README changes - or just respond here and I'll make the updates.

**Comment by westernheld** [comment]
I am using AMD card and its not working. Regarding to this Ticket at the docker project (https://github.com/docker/cli/issues/2063) the issue is in general an it is still open

**Comment by Thermi** [comment]
I got it working with another container and described the process here: https://blog.thermicorp.de/?p=113

TL;DR: Install AMDGPU PRO opencl library and forward /dev/dri/card* and associated /Dev/dri/renderD* device. That's it.

**Comment by mharsch** [comment]
There are currently 2 supported OpenCL stacks for AMD GPUs on LInux: amdgpu-pro and ROCm.  The amdgpu-pro stack requires the proprietary drivers be downloaded and installed onto the host per the instructions [here](https://amdgpu-install.readthedocs.io/en/latest/).  The ROCm stack can use either upstream kernel drivers (e.g. Linux 5.3) or DKMS packages provided by AMD as described [here](https://rocmdocs.amd.com/en/latest/Installation_Guide/Installation-Guide.html).  Both stacks ship their own userland OpenCL library.

**Comment by mharsch** [comment]
[here](https://github.com/mharsch/containers/tree/rocm/fah-rocm) is one way to achieve a working ROCm container using this pattern.  Tested using Ubuntu 18.04.4 (5.3 hwe kernel) plus ROCm 3.3.0 DKMS drivers installed on the host (AMD RX Vega 64 GPU).

**Comment by mharsch** [comment]
and [here](https://github.com/mharsch/containers/tree/amd-pro/fah-amd-pro) is an example of a working container using the AMDGPU-PRO stack.  Tested on a host running Ubuntu 18.04.4 (5.3 hwe kernel) plus AMDGPU-PRO drivers (version amdgpu-pro-20.10-1048554-ubuntu-18.04).

**Comment by beberg** [comment]
@mharsch OK, the driver and docker commands are clear enough to add ROCm and AMDGPU sections. AMD is now preferring ROCm?

The goal is one container that works on all GPUs, (and some machines have a mix) so we'll need the changes to make to add to the Dockerfile to add AMD support while maintaining NVIDIA support. If ROCm and AMDGPU conflict we'll have to pick one.

**Comment by mharsch** [comment]
ROCm is the preferred stack for GPU compute on Linux.  It is actively being developed (in the open) and should out-perform AMDGPU-PRO on modern hardware.  That said, there are reasons to support AMDGPU-PRO: more (older) devices are supported, and it has a longer established track record with F@H (e.g. ROCm had issues with core21 and only started being stable as of core22).  Also, Linux workstations running graphical environments (especially 3D) will be running AMDGPU-PRO, as ROCm seems to be more focused on headless server type (compute-focused) machines.

Universal GPU support is a nice idea, but wouldn't it be best achieved with separate images for each GPU stack (e.g. fah-nvidia, fah-rocm, etc)?  Otherwise, we'll have to include all the userspace contents for each stack and ensure they play nice together for all combinations (e.g. multiple OpenCL implementations).  Also, the NVIDIA case requires nvida-docker be installed on the docker host (which itself requires the NVIDIA drivers be installed).  Would AMD-only users also have to install the NVIDIA stack requirements in order to launch the one container?  At the least, it would certainly require test coverage.

**Comment by Thermi** [comment]
Detecting it should be doable by checking the list of loaded kernel modules and moving the opencl libs as necessary in the container FS.

**Comment by beberg** [comment]
Assuming ROCm is still the choice going forward I'd like to get an AMD config started. 

@mharsch would you like to update to 7.6.21 and re-patch README your https://github.com/mharsch/containers/tree/rocm/fah-rocm and create a PR into containers/fah-gpu-amd/  ?

**Comment by beberg** [comment]
fah-gpu-amd container support merged, thanks to @mharsch ! I will work on setting up the foldingathome/fah-gpu-amd container.
